### ISSUE TITLE
Refactor: update `QuantumServerless.__init__` to accept list of `Provider`s

### ISSUE THREAD

**Issue by psschwei** [Qiskit-Extensions/quantum-serverless#273]
### Steps to reproduce the problem

* Launch a kind cluster
* Install quantum-serverless via helm chart (disabling keycloak for the time being)
* Run the quickstart example, using a kuberay provider instead of a docker one

```python
$ python
Python 3.9.16 (main, Dec  7 2022, 00:00:00) 
[GCC 12.2.1 20221121 (Red Hat 12.2.1-4)] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> from quantum_serverless import QuantumServerless, Program
>>> from quantum_serverless.core.state import RedisStateHandler
>>> 
>>> serverless = QuantumServerless({
...     "providers": [{
...         "name": "kuberay-kind",
...         "host": "http://localhost:31888",
...         "compute_resource": {
...             "name": "kuberay-resource",
...             "resources": {
...                 "cpu": 2, 
...                 "memory": 4, 
...                 "worker_replicas": 1
...             }
...         }
...     }]
... })
>>> serverless.set_provider("kuberay-kind")
<QuantumServerless | providers [local, kuberay-kind]>
>>> state_handler = RedisStateHandler("localhost", 6379)
>>> 
>>> # create out program
>>> program = Program(
...     name="my_program",
...     entrypoint="program.py" # set entrypoint as our program.py file
... )
>>> job = serverless.run_program(program)
WARNING:root:Job has not been submitted as no provider with remote host has been configured. Selected provider: <Provider: kuberay-kind>
>>> 
```

### What is the current behavior?

When running the program, we get a `WARNING:root:Job has not been submitted as no provider with remote host has been configured. Selected provider: <Provider: kuberay-kind>` message 

### What is the expected behavior?

Job is submitted successfully

**Comment by psschwei** [comment]
If we run `serverless.provider()` instead of `.set_provider()`, we get another error:

```python
$ python
Python 3.9.16 (main, Dec  7 2022, 00:00:00) 
[GCC 12.2.1 20221121 (Red Hat 12.2.1-4)] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> from quantum_serverless import QuantumServerless, Program
>>> from quantum_serverless.core.state import RedisStateHandler
>>> 
>>> serverless = QuantumServerless({
...     "providers": [{
...         "name": "kuberay-kind",
...         "host": "localhost:31888",
...         "compute_resource": {
...             "name": "kuberay-resource",
...             "resources": {
...                 "cpu": 2, 
...                 "memory": 4, 
...                 "worker_replicas": 1
...             }
...         }
...     }]
... })
>>> serverless.provider("kuberay-kind") 
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/paulschw/quantum/quantum-serverless/venv/lib64/python3.9/site-packages/quantum_serverless/quantum_serverless.py", line 265, in provider
    return self.context(provider=provider, **kwargs)
  File "/home/paulschw/quantum/quantum-serverless/venv/lib64/python3.9/site-packages/quantum_serverless/quantum_serverless.py", line 248, in context
    return provider.context(**kwargs)
  File "/home/paulschw/quantum/quantum-serverless/venv/lib64/python3.9/site-packages/quantum_serverless/core/provider.py", line 190, in context
    return self.compute_resource.context(**kwargs)
  File "/home/paulschw/quantum/quantum-serverless/venv/lib64/python3.9/site-packages/quantum_serverless/core/provider.py", line 103, in context
    return ray.init(**init_args)
  File "/home/paulschw/quantum/quantum-serverless/venv/lib64/python3.9/site-packages/ray/_private/client_mode_hook.py", line 105, in wrapper
    return func(*args, **kwargs)
  File "/home/paulschw/quantum/quantum-serverless/venv/lib64/python3.9/site-packages/ray/_private/worker.py", line 1409, in init
    ray_params = ray._private.parameter.RayParams(
  File "/home/paulschw/quantum/quantum-serverless/venv/lib64/python3.9/site-packages/ray/_private/parameter.py", line 243, in __init__
    self._check_usage()
  File "/home/paulschw/quantum/quantum-serverless/venv/lib64/python3.9/site-packages/ray/_private/parameter.py", line 412, in _check_usage
    assert "memory" not in self.resources, build_error("memory", "memory")
AssertionError: {'cpu': 2, 'memory': 4, 'worker_replicas': 1} -> `memory` cannot be a custom resource because it is one of the default resources ({'memory', 'CPU', 'object_store_memory', 'GPU'}). Use `memory` instead. For example, use `ray start --memory=1` instead of `ray start --resources={'memory': 1}`
>>> 
```

**Comment by IceKhan13** [comment]
Need to refactor `QuantumServerless.__init__` method to accept `List[Union[Provider, KuberayProvider]]` and act as nesessary instead of list of jsons

**Comment by psschwei** [comment]
This will be fixed by #274
